### PR TITLE
Deduplicating list data

### PR DESCRIPTION
**Pull request by robertbastian**
This PR splits the list resource key into three keys (and/or/unit), and deduplicates the data that is required for each type of list.

### CODE REVIEW

File: experimental/list_formatter/src/provider.rs [hunk 1]
```diff
@@ -6,20 +6,28 @@
...
 }
 
 /// Symbols and metadata required for [`ListFormatter`](crate::ListFormatter).
+/// Absent values follow this fallback structure:
+/// ", " - start - middle
+///            |-- end - pair
+///            |     \ short_end - short_pair
+///            |               \ narrow_end - narrow_pair
+///             \ short_start - short_middle
+///                         \ narrow_start <- narrow_middle
```
Comment (sffc) [inline experimental/list_formatter/src/provider.rs:30]: Another idea:

```
short_start
    start
    narrow_start
short_middle
    middle
    narrow_middle
short_pair
    pair
    narrow_pair
short_end
    end
    narrow_end
```

But actually I think we should store an `enum` which is either the pattern or a "flat" enum, and the "flat" enum references another field that should be used as fallback

File: experimental/list_formatter/src/provider.rs [hunk 2]
```diff
@@ -6,20 +6,28 @@
...
 
 pub mod key {
     //! Resource keys for [`list_formatter`](crate).
     use icu_provider::{resource_key, ResourceKey};
 
     // Resource key: symbols used for list formatting.
-    pub const LIST_FORMAT_V1: ResourceKey = resource_key!(ListFormatter, "list", 1);
+    pub const LIST_FORMAT_AND_V1: ResourceKey = resource_key!(ListFormatter, "list/and", 1);
+    pub const LIST_FORMAT_OR_V1: ResourceKey = resource_key!(ListFormatter, "list/or", 1);
+    pub const LIST_FORMAT_UNIT_V1: ResourceKey = resource_key!(ListFormatter, "list/unit", 1);
```
Comment (sffc) [inline experimental/list_formatter/src/provider.rs:20]:
```suggestion
    pub const LIST_FORMAT_AND_V1: ResourceKey = resource_key!(ListFormatter, "and", 1);
    pub const LIST_FORMAT_OR_V1: ResourceKey = resource_key!(ListFormatter, "or", 1);
    pub const LIST_FORMAT_UNIT_V1: ResourceKey = resource_key!(ListFormatter, "unit", 1);
```

File: provider/cldr/src/transform/list/mod.rs [hunk 3]
```diff
@@ -57,18 +64,35 @@ impl<'data> DataProvider<'data, ListFormatterPatternsV1Marker> for ListProvider<
     ) -> Result<DataResponse<'data, ListFormatterPatternsV1Marker>, DataError> {
         Self::supports_key(&req.resource_path.key)?;
         let langid = req.try_langid()?;
-        let patterns = match self.data.binary_search_by_key(&langid, |(lid, _)| lid) {
+        let data = match self.data.binary_search_by_key(&langid, |(lid, _)| lid) {
             Ok(idx) => &self.data[idx].1.list_patterns,
             Err(_) => return Err(DataError::MissingResourceOptions(req.clone())),
         };
```
Comment (sffc) [inline provider/cldr/src/transform/list/mod.rs:71]:
```suggestion
match req.resource_path.key {
    key::LIST_FORMAT_AND_V1 => parse_and_patterns(data)
    // ...
}
```

File: provider/cldr/src/transform/list/mod.rs [hunk 4]
```diff
@@ -57,18 +64,35 @@ impl<'data> DataProvider<'data, ListFormatterPatternsV1Marker> for ListProvider<
...
 
+        let patterns = if key::LIST_FORMAT_AND_V1
+            .match_key(req.resource_path.key)
+            .is_ok()
+        {
+            parse_and_patterns(data).unwrap()
```
Comment (sffc) [inline provider/cldr/src/transform/list/mod.rs:76]:
```suggestion
            parse_and_patterns(data).map_err(DataError::from_resc_error)
```

File: provider/cldr/src/transform/list/mod.rs [hunk 5]
```diff
@@ -86,6 +110,9 @@ impl<'data> IterableDataProviderCore for ListProvider<'data> {
         let list: Vec<ResourceOptions> = self
             .data
             .iter()
+            // ur-IN has a buggy pattern ("{1}, {0}") which violates
```
Comment (sffc) [inline provider/cldr/src/transform/list/mod.rs:113]: Request: can we reference a bug here, to remove this workaround eventually?

File: experimental/list_formatter/src/provider.rs (hunk 1 above)
Comment (robertbastian) [reply experimental/list_formatter/src/provider.rs:30]: Will land this and work on a better fallback mechanism in a follow-up

**Review by sffc — APPROVED** [review]

**Pull request merged**